### PR TITLE
feat: Context scoped settings

### PR DESCRIPTION
**Pull request by calavera**
Add functionality to override redirect and header rules per context.

Expose two functions to help gathering all redirect and header rules by context, considering the running scope.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -115,43 +126,206 @@ impl Config {
...
         // Read the env variables from the global "build" context.
-        if let Some(c) = self.build {
-            if let Some(ref env) = c.environment {
-                for (k, v) in env {
-                    result.insert(k.to_string(), v.to_string());
-                }
+        if let Some(env) = self.build.and_then(|b| b.environment) {
+            for (k, v) in env {
+                result.insert(k.to_string(), v.to_string());
+            }
```
Comment (NeoLegends) [inline src/lib.rs]: This looks like we could use `.extend(...)`.

File: src/lib.rs [hunk 2]
```diff
@@ -115,43 +126,206 @@ impl Config {
...
+
+        let context = match &self.context {
+            Some(c) => c,
+            None => return result,
+        };
+
+        if let Some(env) = context.get(ctx).and_then(|x| x.build.environment.as_ref()) {
+            for (k, v) in env {
+                result.insert(k.to_string(), v.to_string());
+            }
```
Comment (NeoLegends) [inline src/lib.rs]: Here as well.

File: src/lib.rs [hunk 3]
```diff
@@ -115,43 +126,206 @@ impl Config {
...
+        }
+
+        if let Some(env) = context
+            .get(branch)
+            .and_then(|x| x.build.environment.as_ref())
+        {
+            for (k, v) in env {
+                result.insert(k.to_string(), v.to_string());
+            }
```
Comment (NeoLegends) [inline src/lib.rs]: Here as well.

File: src/lib.rs [hunk 4]
```diff
@@ -115,43 +126,206 @@ impl Config {
     /// "#;
     ///
     /// let config = netlify_toml::from_str(io).unwrap();
-    /// let env = config.context_env("deploy-preview", "new-styles");
+    /// let env = config.scoped_env("deploy-preview", "new-styles");
     /// ```
-    pub fn context_env(self, ctx: &str, branch: &str) -> HashMap<String, String> {
+    pub fn scoped_env(self, ctx: &str, branch: &str) -> HashMap<String, String> {
         let mut result = HashMap::<String, String>::new();
```
Comment (NeoLegends) [inline src/lib.rs]: Can we leave off the type parameters here and let inference do its thing?

File: src/lib.rs [hunk 5]
```diff
@@ -115,43 +126,206 @@ impl Config {
...
+        let context = match &self.context {
+            Some(c) => c,
+            None => return self.headers.clone(),
+        };
+
+        let mut global_index = IndexMap::<String, Header>::new();
+        if let Some(global) = &self.headers {
+            for r in global {
+                global_index.insert(r.path.clone(), r.clone());
+            }
```
Comment (NeoLegends) [inline src/lib.rs:288]: Here as well.

File: src/lib.rs [hunk 6]
```diff
@@ -115,43 +126,199 @@ impl Config {
...
+        let context = match &self.context {
+            Some(c) => c,
+            None => return self.redirects.clone(),
+        };
+
+        let mut global_index = IndexMap::new();
+        if let Some(global) = &self.redirects {
+            for r in global {
+                global_index.insert(r.from.clone(), r.clone());
             }
```
Comment (mraerino) [inline src/lib.rs:200]:
```suggestion
            global_index.extend(global.iter().map(|r| (r.from.clone(), r.clone())));
```

File: src/lib.rs [hunk 7]
```diff
@@ -115,43 +126,199 @@ impl Config {
...
+                if global_index.contains_key(&r.from) {
+                    global_index.insert(r.from.clone(), r.clone());
+                } else {
+                    let mut swap = IndexMap::<String, Redirect>::new();
+                    swap.insert(r.from.clone(), r.clone());
+                    swap.extend(global_index);
+                    global_index = swap;
                 }
             }
+        };
```
Comment (mraerino) [inline src/lib.rs:227]: this suggestion makes the order of the prepended rules a bit more obvious and only runs the swap once.

```rust
        let mut insert = Vec::new();

        if let Some(ct) = context.get(ctx).and_then(|x| x.redirects.clone()) {
            for r in ct {
                if global_index.contains_key(&r.from) {
                    global_index.insert(r.from.clone(), r);
                } else {
                    insert.push(r);
                }
            }
        };

        if let Some(ct) = context.get(branch).and_then(|x| x.redirects.clone()) {
            for r in ct {
                if global_index.contains_key(&r.from) {
                    global_index.insert(r.from.clone(), r);
                } else {
                    insert.push(r);
                }
            }
        };

        if !insert.is_empty() {
            let mut swap = IndexMap::<String, Redirect>::new();
            swap.extend(insert.into_iter().map(|r| (r.from.clone(), r)));
            swap.extend(global_index);
            global_index = swap;
        }
```

**Review by mraerino — COMMENTED** [review]
`scoped_redirects` and `scoped_headers` are really similar. could you put the logic into a single generic function?

File: src/lib.rs [hunk 8]
```diff
@@ -115,43 +126,199 @@ impl Config {
...
+
+        let context = match &self.context {
+            Some(c) => c,
+            None => return result,
+        };
+
+        context
+            .get(ctx)
+            .and_then(|x| x.build.environment.clone())
+            .and_then(|env| Some(result.extend(env)));
```
Comment (NeoLegends) [inline src/lib.rs:147]:
```suggestion
            .map(|env| result.extend(env));
```

File: src/lib.rs [hunk 9]
```diff
@@ -115,43 +126,199 @@ impl Config {
...
+
+        context
+            .get(branch)
+            .and_then(|x| x.build.environment.clone())
+            .and_then(|env| Some(result.extend(env)));
```
Comment (NeoLegends) [inline src/lib.rs:152]:
```suggestion
            .map(|env| result.extend(env));
```

File: src/lib.rs [hunk 10]
```diff
@@ -115,43 +126,199 @@ impl Config {
...
+                    let mut swap = IndexMap::<String, Redirect>::new();
+                    swap.insert(r.from.clone(), r.clone());
+                    swap.extend(global_index);
+                    global_index = swap;
                 }
             }
+        };
+
+        if !global_index.is_empty() {
+            Some(global_index.values().map(|r| r.clone()).collect())
```
Comment (NeoLegends) [inline src/lib.rs:230]:
```suggestion
            Some(global_index.values().cloned().collect())
```

File: src/lib.rs [hunk 11]
```diff
@@ -115,43 +126,199 @@ impl Config {
...
+                    let mut swap = IndexMap::<String, Header>::new();
+                    swap.insert(r.path.clone(), r.clone());
+                    swap.extend(global_index);
+                    global_index = swap;
+                }
+            }
+        };
+
+        if !global_index.is_empty() {
+            Some(global_index.values().map(|r| r.clone()).collect())
```
Comment (NeoLegends) [inline src/lib.rs:318]:
```suggestion
            Some(global_index.values().cloned().collect())
```

**Comment by mraerino** [comment]
it doesn't make sense to introduce that right now. let's wait until we have the routing proxy rolled more widely

**Pull request closed without merging**